File: Beecrowd_Parte4/Quimica_1326.java
import java.util.*;
//saida errada
public class Quimica_1326 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        while (true) {
            int R = sc.nextInt();
            if (R == 0) break;
            sc.nextLine(); // Consumir a linha restante

            Map<String, Integer> inDegree = new HashMap<>();
            Map<String, List<String>> graph = new HashMap<>();
            List<String> reactions = new ArrayList<>();

            for (int i = 0; i < R; ++i) {
                String reaction = sc.nextLine();
                reactions.add(reaction);
                String[] parts = reaction.split(" ");
                String S1 = parts[0];
                String S2 = parts[2];
                String S3 = parts[4];

                graph.putIfAbsent(S1, new ArrayList<>());
                graph.putIfAbsent(S2, new ArrayList<>());
                graph.putIfAbsent(S3, new ArrayList<>());

                graph.get(S1).add(S3);
                graph.get(S2).add(S3);

                inDegree.put(S3, inDegree.getOrDefault(S3, 0) + 2);
                inDegree.putIfAbsent(S1, 0);
                inDegree.putIfAbsent(S2, 0);
            }

            // Encontrar a substância final
            String finalProduct = reactions.get(R - 1).split(" ")[4];

            // Realizar a ordenação topológica
            Queue<String> queue = new LinkedList<>();
            Map<String, Integer> level = new HashMap<>();

            for (String substance : inDegree.keySet()) {
                if (inDegree.get(substance) == 0) {
                    queue.add(substance);
                    level.put(substance, 1);
                }
            }

            while (!queue.isEmpty()) {
                String current = queue.poll();
                int currentLevel = level.get(current);

                for (String neighbor : graph.get(current)) {
                    inDegree.put(neighbor, inDegree.get(neighbor) - 1);
                    if (inDegree.get(neighbor) == 0) {
                        queue.add(neighbor);
                        level.put(neighbor, currentLevel + 1);
                    }
                }
            }

            // O número de recipientes necessários é a profundidade máxima da DAG
            int containersNeeded = level.get(finalProduct);
            System.out.printf("%s requires %d containers%n", finalProduct, containersNeeded);
        }

        sc.close();
    }
}

File: Beecrowd_Parte4/Quimica_1326.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
//saida errada
#define MAX_R 1000
#define MAX_S 100
#define MAX_NAME 50

typedef struct {
    char name[MAX_NAME];
    int inDegree;
    int level;
    int numNeighbors;
    char neighbors[MAX_S][MAX_NAME];
} Substance;

Substance substances[MAX_S];
int numSubstances;

Substance* findSubstance(char* name) {
    for (int i = 0; i < numSubstances; ++i) {
        if (strcmp(substances[i].name, name) == 0) {
            return &substances[i];
        }
    }
    strcpy(substances[numSubstances].name, name);
    return &substances[numSubstances++];
}

void addEdge(Substance* from, Substance* to) {
    strcpy(from->neighbors[from->numNeighbors++], to->name);
    ++to->inDegree;
}

int main() {
    while (1) {
        int R;
        scanf("%d", &R);
        if (R == 0) break;
        getchar();  // Consumir a linha restante

        numSubstances = 0;
        memset(substances, 0, sizeof(substances));

        char finalProduct[MAX_NAME];
        for (int i = 0; i < R; ++i) {
            char S1[MAX_NAME], S2[MAX_NAME], S3[MAX_NAME];
            scanf("%s + %s -> %s", S1, S2, S3);
            getchar();  // Consumir a linha restante

            Substance* substance1 = findSubstance(S1);
            Substance* substance2 = findSubstance(S2);
            Substance* substance3 = findSubstance(S3);

            addEdge(substance1, substance3);
            addEdge(substance2, substance3);

            if (i == R - 1) {
                strcpy(finalProduct, S3);
            }
        }

        // Realizar a ordenação topológica
        Substance* queue[MAX_S];
        int front = 0, back = 0;

        for (int i = 0; i < numSubstances; ++i) {
            if (substances[i].inDegree == 0) {
                queue[back++] = &substances[i];
                substances[i].level = 1;
            }
        }

        while (front != back) {
            Substance* current = queue[front++];
            int currentLevel = current->level;

            for (int i = 0; i < current->numNeighbors; ++i) {
                Substance* neighbor = findSubstance(current->neighbors[i]);
                --neighbor->inDegree;
                if (neighbor->inDegree == 0) {
                    queue[back++] = neighbor;
                    neighbor->level = currentLevel + 1;
                }
            }
        }

        // O número de recipientes necessários é a profundidade máxima da DAG
        int containersNeeded = findSubstance(finalProduct)->level;
        printf("%s requires %d containers\n", finalProduct, containersNeeded);
    }

    return 0;
}